### PR TITLE
raftstore: set term in read index response

### LINKED ISSUE #7353: raft::raft_log::RaftLog<T>::commit_to paniked
### What version of TiKV are you using?

```
/ # /tikv-server -V
TiKV
Release Version:   4.0.0-beta.2
Git Commit Hash:   f6e0bb5b2719b2f1262fc2898e12ff3742c9e332
Git Commit Branch: HEAD
UTC Build Time:    2020-04-02 01:15:12
Rust Version:      rustc 1.42.0-nightly (0de96d37f 2019-12-19)
Enable Features:   jemalloc portable sse protobuf-codec
Profile:           dist_release
```

### What operating system and CPU are you using?

GCP n1-standard-32

### Steps to reproduce

run tipocket/bank with 
```
-tidb-replica-read=leader-and-follower
-nemesis=random_kill,kill_pd_leader_5min,partition_one,subcritical_skews,big_skews,shuffle-leader-scheduler,shuffle-region-scheduler,random-merge-scheduler
```

### What did you expect?

```
NONE
```

### What did happened?

```
2020/04/02 15:51:41 control.go:497: [fatal] 1 panics occurred in ns: tipocket-scbank2 pod tipocket-scbank2-tikv-1. Content: [[2020/04/02 15:45:20.863 +00:00] [FATAL] [lib.rs:480] ["to_commit 177238 is out of range [last_index 177215], raft_id: 2003, region_id: 2001"] [backtrace="stack backtrace:
   0: tikv_util::set_panic_hook::{{closure}}
             at components/tikv_util/src/lib.rs:479
   1: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:475
   2: rust_begin_unwind
             at src/libstd/panicking.rs:375
   3: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:326
   4: raft::raft_log::RaftLog<T>::commit_to
             at tikv/<::std::macros::panic macros>:9
   5: raft::raft_log::RaftLog<T>::maybe_commit
             at root/.cargo/git/checkouts/raft-rs-841f8a6db665c5c0/b9891b6/src/raft_log.rs:429
   6: raft::raft::Raft<T>::step_follower
             at root/.cargo/git/checkouts/raft-rs-841f8a6db665c5c0/b9891b6/src/raft.rs:1851
      raft::raft::Raft<T>::step
             at root/.cargo/git/checkouts/raft-rs-841f8a6db665c5c0/b9891b6/src/raft.rs:1155
   7: raft::raw_node::RawNode<T>::step
             at root/.cargo/git/checkouts/raft-rs-841f8a6db665c5c0/b9891b6/src/raw_node.rs:339
      raftstore::store::peer::Peer::step
             at tikv/components/raftstore/src/store/peer.rs:762
      raftstore::store::fsm::peer::PeerFsmDelegate<T,C>::on_raft_message
             at tikv/components/raftstore/src/store/fsm/peer.rs:894
   8: raftstore::store::fsm::peer::PeerFsmDelegate<T,C>::handle_msgs
             at tikv/components/raftstore/src/store/fsm/peer.rs:282
   9: <raftstore::store::fsm::store::RaftPoller<T,C> as batch_system::batch::PollHandler<raftstore::store::fsm::peer::PeerFsm<engine_rocks::engine::RocksEngine>,raftstore::store::fsm::store::StoreFsm>>::handle_normal
             at tikv/components/raftstore/src/store/fsm/store.rs:702
  10: batch_system::batch::Poller<N,C,Handler>::poll
             at tikv/components/batch-system/src/batch.rs:290
  11: batch_system::batch::BatchSystem<N,C>::spawn::{{closure}}
             at tikv/components/batch-system/src/batch.rs:394
      std::sys_common::backtrace::__rust_begin_short_backtrace
             at rustc/0de96d37fbcc54978458c18f5067cd9817669bc8/src/libstd/sys_common/backtrace.rs:136
  12: std::thread::Builder::spawn_unchecked::{{closure}}::{{closure}}
             at rustc/0de96d37fbcc54978458c18f5067cd9817669bc8/src/libstd/thread/mod.rs:469
      <std::panic::AssertUnwindSafe<F> as core::ops::function::FnOnce<()>>::call_once
             at rustc/0de96d37fbcc54978458c18f5067cd9817669bc8/src/libstd/panic.rs:318
      std::panicking::try::do_call
             at rustc/0de96d37fbcc54978458c18f5067cd9817669bc8/src/libstd/panicking.rs:292
      std::panicking::try
             at rustc/0de96d37fbcc54978458c18f5067cd9817669bc8//src/libpanic_unwind/lib.rs:78
      std::panic::catch_unwind
             at rustc/0de96d37fbcc54978458c18f5067cd9817669bc8/src/libstd/panic.rs:394
      std::thread::Builder::spawn_unchecked::{{closure}}
             at rustc/0de96d37fbcc54978458c18f5067cd9817669bc8/src/libstd/thread/mod.rs:468
      core::ops::function::FnOnce::call_once{{vtable.shim}}
             at rustc/0de96d37fbcc54978458c18f5067cd9817669bc8/src/libcore/ops/function.rs:232
  13: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at rustc/0de96d37fbcc54978458c18f5067cd9817669bc8/src/liballoc/boxed.rs:1022
  14: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at rustc/0de96d37fbcc54978458c18f5067cd9817669bc8/src/liballoc/boxed.rs:1022
      std::sys_common::thread::start_thread
             at src/libstd/sys_common/thread.rs:13
      std::sys::unix::thread::Thread::new::thread_start
             at src/libstd/sys/unix/thread.rs:80
  15: <unknown>
  16: clone
"] [location=/root/.cargo/git/checkouts/raft-rs-841f8a6db665c5c0/b9891b6/src/raft_log.rs:237] [thread_name=raftstore-4-0]
```

[tipocket-scbank2-p7v8c.log](https://github.com/tikv/tikv/files/4424713/tipocket-scbank2-p7v8c.log)

### PR DESCRIPTION
**Pull request by gengliqi**
Signed-off-by: Liqi Geng <gengliqiii@gmail.com>

### What problem does this PR solve?

close #7353 

Problem Summary:

TiKV may panic in raft-rs because of tikv/raft-rs#346 and the subtle hack code.
Some details in https://github.com/tikv/tikv/issues/7353#issuecomment-608247067

### Related changes

- Need to cherry-pick to the release branch

### Check List 

Tests 

- Integration test

Side effects

No

### CODE REVIEW

File: tests/integrations/raftstore/test_replica_read.rs [hunk 1]
```diff
@@ -208,33 +240,43 @@ fn test_read_hibernated_region() {
...
+        .add_recv_filter(3, filter);
+    cluster.must_put(b"k2", b"v2");
+    let resp1_ch = async_read_on_peer(&mut cluster, peer_on_store3.clone(), region.clone(), b"k2", true, true);
+    // must be timeout
+    assert!(resp1_ch.recv_timeout(Duration::from_micros(100)).is_err());
+    
+    cluster.sim.wl().clear_recv_filters(3);
+
+    cluster.must_put(b"k3", b"v3");
+    let resp2_ch = async_read_on_peer(&mut cluster, peer_on_store3, region, b"k2", true, true);
```
Comment (BusyJay) [inline tests/integrations/raftstore/test_replica_read.rs:279]: I think you can just wait for `resp1_ch`.

File: components/raftstore/src/store/peer.rs [hunk 2]
```diff
@@ -744,6 +744,7 @@ impl Peer {
             if let LeaseState::Valid = state {
                 let mut resp = eraftpb::Message::default();
                 resp.set_msg_type(MessageType::MsgReadIndexResp);
+                resp.term = self.term();
```
Comment (BusyJay) [inline components/raftstore/src/store/peer.rs:747]: Better check if all required fields are set. I can see there are still places that construct raft messages by ourselves, need to be checked too.

Comment (gengliqi) [reply components/raftstore/src/store/peer.rs:747]: Another raft message constructed by ourselves is `transfer_leader` but it is local msg for raft-rs.In TiKV we send it for checking transfer leader condition and do not use term field so it's ok not to set it.

File: tests/integrations/raftstore/test_replica_read.rs (hunk 1 above)
Comment (gengliqi) [reply tests/integrations/raftstore/test_replica_read.rs]: Done

File: tests/integrations/raftstore/test_replica_read.rs [hunk 3]
```diff
@@ -208,33 +240,40 @@ fn test_read_hibernated_region() {
     assert!(!resp2.get_header().has_error(), "{:?}", resp2);
 }
 
-#[derive(Default)]
-struct CommitToFilter {
-    // map[peer_id] -> committed index.
-    committed: Arc<Mutex<HashMap<u64, u64>>>,
-}
+#[test]
+fn test_replica_read_on_stale_peer() {
```
Comment (BusyJay) [inline tests/integrations/raftstore/test_replica_read.rs:244]: I don't think people can guess what it's actually testing, better add some comment.

Comment (gengliqi) [reply tests/integrations/raftstore/test_replica_read.rs:247]: Done

**Pull request merged**